### PR TITLE
Fix bugs in XQSuite and XQuery tests written for XQSuite

### PR DESCRIPTION
**Pull request by adamretter**
Closes https://github.com/eXist-db/exist/issues/4612

Unfortunately fixing a number of bugs in XQSuite has revealed:
1. A number of tests written for XQSuite that are invalid XQuery (according to the W3C spec)
2. A number of sub-type promotion errors in eXist-db

### CODE REVIEW

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql [hunk 1]
```diff
@@ -526,47 +527,135 @@ declare %private function test:map-named-argument($testArgs as element(value)*,
...
+            return
+                xs:dateTime($varg)
+
+            case "xs:time"
+            return
+                xs:time($varg)
+
+            case "element()"
+            return
+                fn:parse-xml($varg)/element()
```
Comment (joewiz) [inline exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:646]: The `fn:parse-xml` function returns a `document-node`, so selecting `/element()` here will return an empty sequence. See https://www.w3.org/TR/xpath-functions-31/#func-parse-xml.

Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:646]: I think this should work as I usually select the first child node() when dealing with document-nodes

Comment (joewiz) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:646]: Ah, you're quite right. Please disregard my comment above.

**Comment by line-o** [comment]
Please have a look at the test failures.

**Review by dizzzz — APPROVED** [review]

**Comment by adamretter** [comment]
> Please have a look at the test failures
@line-o All tests are now passing.

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql [hunk 2]
```diff
@@ -801,7 +801,7 @@ declare %private function test:assertEqualsPermutation($values as element(value)
             let $tests :=
                 for $item in $result
                 return
-                    some $value in $values satisfies test:equals($item, $value)
+                    some $value in $values satisfies test:equals($value, $item)
```
Comment (joewiz) [inline exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:804]: Good catch! We should consider renaming `$item` to `$result` in this function; `$result` to `$results`; and perhaps even `$value` and `$values` to `$annotation-value` and `$annotation-values`. This would make it quite apparent how these variables need to align with the `test:equals()` function signature: `test:equals($annotation-value as element(value), $result as item())`. (Perhaps in a future PR, that is.)

Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:804]: I would be in favour of changing the name as part of this PR

**Comment by adamretter** [comment]
> What about the error in Deploy / Build and Test Images?

I don't know what the Bats tests do and I don't understand them. Any idea who added them and/or what they do?

**Comment by duncdrum** [comment]
They test the docker container. In this case https://github.com/eXist-db/exist/actions/runs/3574037174/jobs/6008812699#step:11:12 if the logs are free of errors on startup. They are not in this branch

**Comment by adamretter** [comment]
Okay thanks @duncdrum, do you know where I can see the logs?

**Comment by adamretter** [comment]
@duncdrum Unless I am mistaken the `exist.log` or Docker console log is not available from the CI for this?

**Comment by adamretter** [comment]
I build a Docker image locally, and it does indeed show that there was a hidden XQuery error in Monex which is shown by the fixes in this PR:

```
30 Nov 2022 22:03:45,321 [main] ERROR (AutoDeploymentTrigger.java [execute]:127) - Exception during deployment of app monex-3.0.4.xar: Error found while processing repo.xml: err:XPTY0004 xs:string(/db/apps/monex/data) is not a sub-type of xs:anyURI [at line 12, column 15, source: /exist/etc/../data/expathrepo/monex-3.0.4/post-install.xql]
org.expath.pkg.repo.PackageException: Error found while processing repo.xml: err:XPTY0004 xs:string(/db/apps/monex/data) is not a sub-type of xs:anyURI [at line 12, column 15, source: /exist/etc/../data/expathrepo/monex-3.0.4/post-install.xql]
	at org.exist.repo.Deployment.deploy(Deployment.java:469) ~[exist.uber.jar:6.1.0-SNAPSHOT]
	at org.exist.repo.Deployment.installAndDeploy(Deployment.java:281) ~[exist.uber.jar:6.1.0-SNAPSHOT]
```

Seems we will also need to fix Monex and get a new release of that out...

**Comment by adamretter** [comment]
Okay I fixed this issue in a quick hotfix of Monex 3.0.5 and this has fixed the build here.

**Review by line-o — COMMENTED** [review]
Since one of the bundled applications seems to choke on the changes in this PR we need to investigate first if other packages might also be affected.

**Comment by line-o** [comment]
related to #4632

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql [hunk 3]
```diff
@@ -526,47 +527,135 @@ declare %private function test:map-named-argument($testArgs as element(value)*,
 (:~
  : For %args: transform each annotation parameter into the type required for the function parameter.
  :)
-declare %private function test:map-arguments($testArgs as xs:string*, $funcArgs as element(argument)*) {
-    if (exists($testArgs)) then
-        if (count($testArgs) != count($funcArgs)) then
-            error(
+declare %private function test:map-arguments($test-args as element(value)*, $func-args as element(argument)*) {
+    if (fn:not(fn:empty($test-args))) then
```
[the remark this answers is not in the document]
Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:531]: and in all following instances

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql [hunk 4]
```diff
@@ -706,13 +795,13 @@ declare %private function test:assertEquals($values as item()*, $result as item(
  : Check for equality of the function's result with the value in the annotation.
  : This function transforms the result to a string before checking for equality.
  :)
-declare %private function test:assertEqualsPermutation($values as item()*, $result as item()*) as element(report)? {
-    if (exists($values)) then
-        if (count($values) eq count($result)) then
+declare %private function test:assertEqualsPermutation($values as element(value)*, $result as item()*) as element(report)? {
+    if (fn:not(fn:empty($values))) then
```
Comment (line-o) [inline exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:799]: why is `exists()` not sufficient here?

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql [hunk 5]
```diff
@@ -819,14 +909,27 @@ declare %private function test:assertTrue($result as item()*) as element(report)
...
 declare %private function test:assertError($value as xs:string, $result as item()*) as element(report)? {
-    if ($result) then
-        ()
-    else
-        <report>
-            <failure message="assertError failed. Expected error {$value}"
-                type="failure-error-code-1"/>
-            <output>{ $result }</output>
-        </report>
+    let $ebv :=
```
Comment (line-o) [inline exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:912]: This code looks like it could be written in a more readable way.

File: exist-core/src/test/xquery/xqsuite/xqsuite-assertions-inline.xqm [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+ : Inline Tests for XQSuite assertions.
+ :
+ : @author Adam Retter
+ :)
+module namespace xqtai = "http://exist-db.org/xquery/xqsuite/test/assertions-inline";
+
+(:~
+ : Resource functions to test through XQSuite.
+ :)
+import module namespace xqtares = "http://exist-db.org/xquery/xqsuite/test/assertions/resources"
```
Comment (line-o) [inline exist-core/src/test/xquery/xqsuite/xqsuite-assertions-inline.xqm:45]: Here it is imported and not just the namespace ignored.

File: exist-core/src/test/xquery/xquery3/parseIetfDate.xqm [hunk 7]
```diff
@@ -51,8 +51,10 @@ declare
     %test:args('') %test:assertError("FORG0010")
     %test:args('asdfa') %test:assertError("FORG0010")
 
-    %test:args(0) %test:assertError("FORG0010")
-    %test:args(1) %test:assertError("FORG0010")
+    %test:args(0) %test:assertError("XPTY0004")
```
Comment (line-o) [inline exist-core/src/test/xquery/xquery3/parseIetfDate.xqm:54]: These test cases that will result in XPTY0004 can be removed. They just test the type system of exist and not the function itself.

File: exist-core/src/test/xquery/xquery3/serialize.xql [hunk 8]
```diff
@@ -234,7 +234,7 @@ function ser:serialize-empty-sequence() {
 
 declare
     %test:args(1234)
-    %test:assertEquals('"1234"')
+    %test:assertEquals(1234)
```
Comment (line-o) [inline exist-core/src/test/xquery/xquery3/serialize.xql:237]: Adaptive serialisation will always output a string. The test should reflect that.

File: exist-core/src/test/xquery/xquery3/serialize.xql [hunk 9]
```diff
@@ -243,7 +243,7 @@ function ser:adaptive-simple-atomic($atomic as xs:anyAtomicType) {
...
```
Comment (line-o) [inline exist-core/src/test/xquery/xquery3/serialize.xql:246]: see above

**Review by line-o — DISMISSED** [review]
Please have a look at the comments.

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 5 above)
Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:912]: In which cases does effectiveBoolean value fail - and could that not be mitigated by a call to `exists()` ?

**Comment by line-o** [comment]
Regarding problems with bundled applications: the necessity of commit a1a35e041d79d717f66729bfdf085df258ebcb96 indicates that there might be more problems after this PR is merged.

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 5 above)
Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:912]:
```
declare %private function test:assertError($value as xs:string, $result as item()*) as element(report)? {
    if (exists($result)) then ()
    else
        <report>
            <failure message="assertError failed. Expected error {$value}"
                type="failure-error-code-1"/>
            <output>{ $result }</output>
        </report>
};
```
is sufficient. I just tested it. This change alone shows that the array:put test actually fails. Nothing else needed.

**Comment by line-o** [comment]
This PR contains several - welcome and good - changes that would warrant their own separate PRs. The risk is that we accidentally create more harm than we solve.

**Comment by adamretter** [comment]
> Since one of the bundled applications seems to choke on the changes in this PR we need to investigate first if other packages might also be affected.

@line-o I am afraid that is not specific to this PR. That is just a general statement that: whenever we modify eXist-db's XQuery engine to be more spec compliant, then it is likely that any code that relied on behaviour that is not spec compliant will need to be fixed. This has been true for every XQuery change to eXist-db, and does not relate specifically to this PR.

This change, just like any other PR that changes eXist-db's XQuery compliance in a non-compatible way, is scheduled for a major release.

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql [hunk 10]
```diff
@@ -511,12 +512,12 @@ declare %private function test:map-named-arguments($funcArgs as element(argument
...
 
-declare %private function test:map-named-argument($testArgs as element(value)*, $funcArg as element(argument)) {
+declare %private function test:map-named-argument($test-args as element(value)*, $func-arg as element(argument)) {
     let $data :=
         try {
-            test:cast($testArgs, $funcArg)
+            test:cast($test-args, $func-arg)
         } catch * {
-            error($test:TYPE_ERROR, "Failed to cast annotation arguments to required target type " || $funcArg/@type)
+            fn:error($test:TYPE_ERROR, "Failed to cast annotation arguments to required target type " || $func-arg/@type)
```
[the remark this answers is not in the document]
Comment (adamretter) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:520]: Why? What do we gain for doing that?

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 3 above)
[the remark this answers is not in the document]
Comment (adamretter) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:531]: Why? What do we gain for doing that?

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql [hunk 11]
```diff
@@ -526,47 +527,135 @@ declare %private function test:map-named-argument($testArgs as element(value)*,
...
-                xs:time($targ)
-            case "element()" return
-                parse-xml($targ)/*
-            case "text()" return
-                text { string($targ) }
-            default return
-                $targ
+            case "xs:string"
+            return
+                if ($varg instance of xs:string)
```
[the remark this answers is not in the document]
Comment (adamretter) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:590]: > Isn't this already being handled in test:xdm-value-from-annotation-value?

No it is not.

These are an optimisation, no need to attempt costly conversion if the type is already of the correct type.

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 4 above)
Comment (adamretter) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql]: I changed it back. There was a mix of both styles already present in `xqsuite.xql` before even my changes

File: exist-core/src/test/xquery/xquery3/serialize.xql (hunk 8 above)
Comment (adamretter) [reply exist-core/src/test/xquery/xquery3/serialize.xql:237]: This is not related to adaptive serialization.

This is related to the XQSuite use of XQuery annotations. The XQuery 3.1 spec says that an `Annotation` can have zero or more [Literal](https://www.w3.org/TR/xquery-31/#prod-xquery31-Literal) values. Further, it says that a `Literal` is either a `StringLiteral`, `NumericLiteral`, `IntegerLiteral`, `DecimalLiteral` or `DoubleLiteral`.

If you `assertEquals(1234)` and your test returns an integer, then the integers will now be correctly compared. There was some support for this before in XQSuite but it was broken/incomplete.

File: exist-core/src/test/xquery/xquery3/serialize.xql (hunk 9 above)
Comment (adamretter) [reply exist-core/src/test/xquery/xquery3/serialize.xql:246]: As above - this is not related to adaptive serialization.

File: extensions/modules/mail/src/main/java/org/exist/xquery/modules/mail/MailStoreFunctions.java [hunk 12]
```diff
@@ -52,7 +52,7 @@
  *
  * @see org.exist.xquery.BasicFunction#BasicFunction(org.exist.xquery.XQueryContext, org.exist.xquery.FunctionSignature)
  */
-public class MailStoreFunctions extends BasicFunction
+public class 	MailStoreFunctions extends BasicFunction
```
[the remark this answers is not in the document]
Comment (adamretter) [reply extensions/modules/mail/src/main/java/org/exist/xquery/modules/mail/MailStoreFunctions.java]: Done.

File: exist-core/src/test/xquery/xqsuite/xqsuite-assertions-dynamic.xqm [hunk 13]
```diff
@@ -0,0 +1,234 @@
...
+ : Dynamic Tests for XQSuite assertions.
+ :
+ : @author Adam Retter
+ :)
+module namespace xqtad = "http://exist-db.org/xquery/xqsuite/test/assertions-dynamic";
+
+(:~
+ : Resource functions to test through XQSuite.
+ :)
+declare namespace xqtares = "http://exist-db.org/xquery/xqsuite/test/assertions/resources";
```
[the remark this answers is not in the document]
Comment (adamretter) [reply exist-core/src/test/xquery/xqsuite/xqsuite-assertions-dynamic.xqm:45]: @line-o They do different things. The hint is in the filename!

In the dynamic one we only declare a namespace so we can later *dynamically* import a module of that namespace using `fn:load-xquery-module`. If I instead used `import module` then the module would already be imported statically and I would not be able to import it dynamically later!

File: exist-core/src/test/xquery/xqsuite/xqsuite-assertions-inline.xqm (hunk 6 above)
Comment (adamretter) [reply exist-core/src/test/xquery/xqsuite/xqsuite-assertions-inline.xqm:45]: I am afraid I don't understand your comment. Can you rephrase/elaborate please?

**Comment by adamretter** [comment]
> This PR contains several - welcome and good - changes that would warrant their own separate PRs

As the author of the PR who spent weeks chasing and trying to resolve this issue, an issue that went undetected for years, I strongly disagree. All of the fixes and in this PR are intimately inter-related to each other.

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 5 above)
Comment (adamretter) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:912]: `$result` is of type `item()*`, so in the previous code:

```
if ($result)
```

This is getting the *effective boolean value* of `$result`. That is not the same as your proposed `fn:exists($result)` which would instead just tell you if the sequence is non-empty.

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 10 above)
[the remark this answers is not in the document]
Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:520]: Clarity. It's unnnecessary clutter.

[the remark this answers is not in the document]
Comment (adamretter) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:520]: I think that is subjective, personally I have been writing all XQuery code for sometime now using the `fn` prefix as we have found here that it increases clarity rather than reducing it.

[the remark this answers is not in the document]
Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:520]: I will put it on the agenda for the next community meeting then.

File: exist-core/src/test/xquery/xquery3/parseIetfDate.xqm (hunk 7 above)
Comment (line-o) [reply exist-core/src/test/xquery/xquery3/parseIetfDate.xqm:54]: I believe this issue is now tested elsewhere enough.

File: exist-core/src/test/xquery/xquery3/serialize.xql (hunk 8 above)
Comment (line-o) [reply exist-core/src/test/xquery/xquery3/serialize.xql:237]: Sorry but I have to disagree, @adamretter. This test is for adaptive serialization, if it does _not_ return a string than that is a problem. A reason to reject this PR. So please change it to a string so that we can see that it does in fact return a string and compares them correctly.

File: exist-core/src/test/xquery/xquery3/serialize.xql (hunk 9 above)
Comment (line-o) [reply exist-core/src/test/xquery/xquery3/serialize.xql:246]: We have to be able to test that a string is returned. If this happens to be castable to an integer value should not affect the assertion.

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 10 above)
[the remark this answers is not in the document]
Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:520]: Yeah, we need a definitive way forward. The entire codebase does not use the prefix so it is necessary to decide whether it is used _everywhere_ in the future or _never_.

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 5 above)
Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:912]: OK, I understand now. In that case it would help future readers of the source code to extract the evaluation of a safe effective boolean value into a properly named function.

```xquery
declare %private function test:safeEffectiveBooleanValue ($result as item()*) as xs:boolean {
    try {
        not(not(($result)))
    }
    catch * {
        false()
    }
};

```

File: exist-core/src/test/xquery/xquery3/serialize.xql (hunk 8 above)
Comment (adamretter) [reply exist-core/src/test/xquery/xquery3/serialize.xql:237]: @line-o It may itself be a test for adaptive serialization, but after that the assert-equals comes into play

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 10 above)
[the remark this answers is not in the document]
Comment (adamretter) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:520]: > The entire codebase does not use the prefix

If you grep for `fn:` you will find it in plenty of XQuery files within eXist-db!

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 11 above)
[the remark this answers is not in the document]
Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:590]: How costly is this conversion from string to string in contrast to type check or type check + conversion?
Is this something we have hard numbers on?

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 10 above)
[the remark this answers is not in the document]
Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:520]: The status quo is troubling, I agree.
This PR adds prefixes to existing lines of codes that are otherwise unchanged and the use of prefixes is consistent within the module itself.

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 11 above)
[the remark this answers is not in the document]
Comment (line-o) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:590]: No answer for a week. Will have to be checked afterwards, then.

**Comment by joewiz** [comment]
@adamretter Could you confirm this is a breaking change and needs to wait for 7.0.0?

**Comment by adamretter** [comment]
> Could you confirm this is a breaking change and needs to wait for 7.0.0?

@joewiz Yes, this is a breaking change and requires 7.0.0.

**Comment by adamretter** [comment]
> Why was the hot fix release for money necessary when this code is marked for 7.0.0? Why the rush?

@line-o Without the hotfix to Monex, Monex would not run on 7.0.0-SNAPSHOT. As several people are already using 7.0.0-SNAPSHOT builds and requested Monex... I fixed Monex to work with eXist-db 6 and 7.

File: exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql (hunk 11 above)
[the remark this answers is not in the document]
Comment (adamretter) [reply exist-core/src/main/resources/org/exist/xquery/lib/xqsuite/xqsuite.xql:590]: @line-o I assumed that if you were interested, you could benchmark it.

**Comment by line-o** [comment]
> Without the hotfix to Monex, Monex would not run on 7.0.0-SNAPSHOT. As several people are already using 7.0.0-SNAPSHOT builds and requested Monex... I fixed Monex to work with eXist-db 6 and 7.

Surprising news! Where is the branch for the 7.0.0-SNAPSHOT that is used by several people?

File: exist-core/src/test/xquery/xquery3/serialize.xql (hunk 8 above)
Comment (line-o) [reply exist-core/src/test/xquery/xquery3/serialize.xql:237]: Which should ensure that the expected value _and_ type is returned.

**Review by reinhapa — APPROVED** [review]

**Review by line-o — COMMENTED** [review]
There is still an open question on the adaptive serialization test where the expected value changed from a string to a number literal.
I would like to have proof that XQsuite is able to assert a string '1234' was returned.

**Comment by adamretter** [comment]
> There is still an open question on the adaptive serialization test where the expected value changed from a string to a number literal.

As I tried to explain above there is a sequence of operations that happens. This test is not only testing that a specific string is returned from `fn:serialize` for Adaptive Serialization, but it is also testing the steps that happen after that.

1. A Literal *un-quoted* annotation with value `1234` is passed as the argument to the function, and XQSuite has to parse this and cast it to the correct sub-type of `xs:anyAtomicType`. In this test it will be parsed and cast to `xs:integer`.
2. The function under test is executed by XQSuite.
3.  The function under test eventually calls `fn:serialize` with the Adaptive Serialization mode set; `fn:serialize` returns an `xs:string`.
4. The function under test returns the `xs:string` value.
5. XQSuite needs to convert that `xs:string` value to a type and/or value that is appropriate for comparison to the expected value instructed by the `%test:assert`* annotation.
6. For this test the assert annotation has an *un-quoted` value of `1234`. XQSuite has to coerce this and the result of the function to a type that it can compare according to the XQuery specs for type promotion and type derivation. In this test it will be parsed and cast to `xs:integer`.
7. Finally the two `xs:integer` values are compared for equality.

I hope that helps explain the purpose of this test (which I think is the thing that is not being understood here). To summarise - it covers both the `xs:string` case that you are concerned about, as well as the actual `xs:integer` case that I am trying to test.

> I would like to have proof that XQsuite is able to assert a string '1234' was returned.

That is already present on that test, both implicitly through the explanation above that uses:
```xquery
%test:args(1234)
%test:assertEquals(1234)
```

and also explicitly by the additional assertions on the same function under test:
```xquery
%test:args('Hello "world"!')
%test:assertEquals('"Hello ""world""!"')
```

Additionally... I think you are already aware of many tests, including some written by yourself, within the full XQSuite of tests that already cover that use-case.

**Comment by line-o** [comment]
The two tests in question do test the raw output of serialize (they are part of the spec of our implementation of `fn:serialize`). The function under test _is_ `fn:serialize`. Thus _have to_ return `xs:string` correct?

```xquery
declare
    %test:args(1234)
    %test:assertEquals(1234)
    %test:args('Hello "world"!')
    %test:assertEquals('"Hello ""world""!"')
function ser:adaptive-simple-atomic($atomic as xs:anyAtomicType) {
    ser:adaptive($atomic)
};

declare
    %test:args(1234)
    %test:assertEquals(1234)
    %test:args('Hello "world"!')
    %test:assertEquals('"Hello ""world""!"')
function ser:adaptive-simple-atomic-map-params($atomic as xs:anyAtomicType) {
    ser:adaptive-map-params($atomic)
};
```

**Comment by adamretter** [comment]
> Thus have to return xs:string correct?

I am afraid you are still missing the point of my changes.
Yes, `fn:serialize` returns an `xs:string`, I am not disputing that, I even said above:

> 3. The function under test eventually calls fn:serialize with the Adaptive Serialization mode set; fn:serialize returns an xs:string.

However, that's not the point of these additions that I have made. As I said above:

> > To summarise - it covers both the xs:string case that you are concerned about, as well as the actual xs:integer case that I am trying to test.

Can we please move on?

**Comment by line-o** [comment]
>  actual xs:integer case that I am trying to test.

The test are for `fn:serialize`! They _must_ stay testing that the _return value_ is `xs:string`, otherwise they do not make sense for anyone looking at them in a few months. They were changed in this PR for no apparent reason. Please let's move on by reverting the changes to those two tests.
If you want test type promotion capabilities of XQSuite,  please add them separately to an xqsuite specific testsuite.

**Comment by adamretter** [comment]
@dizzzz @reinhapa Would you be able to spare some time to help @line-o understand what I am trying to achieve here please?

**Comment by reinhapa** [comment]
> > actual xs:integer case that I am trying to test.
> 
> The test are for `fn:serialize`! They _must_ stay testing that the _return value_ is `xs:string`, otherwise they do not make sense for anyone looking at them in a few months. They were changed in this PR for no apparent reason. Please let's move on by reverting the changes to those two tests. If you want test type promotion capabilities of XQSuite, please add them separately to an xqsuite specific testsuite.

@line-o I can not see no problem with the changes from @adamretter  here. In my opinion the fix is as much as important as the test. As I understand it your requirements where covered as explained by Adam earlier. I see no point of changing again.

**Comment by line-o** [comment]
@reinhapa a test that has to ensure the function under test returns the expected value and type.
Since the modified tests test `fn:serialize` it has to be `xs:string`. At the moment this is the case. In this PR the expected return type is changed to a literal `xs:decimal` (`"1234"` was changed to `1234`).
If this does not make a difference technologically it does change the meaning of the test semantically. 
In worst case it can lead to uncaught regressions.
